PWG/EMCAL/EMCALtasks/AliEmcalCorrectionCellTrackMatcher.cxx

```diff
@@ -0,0 +1,252 @@
+  fEmcalTracks(0),
+  fCellTrackMatchdEtadPhi(0),
+  fCellNTrackMatch(0),
+  fCellTrackMatchEbefore(0),
+  fCellTrackMatchEafter(0)
+{
+}
+
+/**
+ * Destructor
+ */
+AliEmcalCorrectionCellTrackMatcher::~AliEmcalCorrectionCellTrackMatcher()
+{
+}
+
+/**
+ * Initialize and configure the component.
+ */
+Bool_t AliEmcalCorrectionCellTrackMatcher::Initialize()
+{
+  // Initialization
+  AliEmcalCorrectionComponent::Initialize();
+
+  GetProperty("mipE", fMipE);
+
+  return kTRUE;
+}
+
+/**
+ * Create run-independent objects for output. Called before running over events.
+ */
+void AliEmcalCorrectionCellTrackMatcher::UserCreateOutputObjects()
+{
+  AliEmcalCorrectionComponent::UserCreateOutputObjects();
+
+  // Determine all particle container array names for naming the AliEmcalParticle array
+  std::string particleContainerNames = "";
+  bool firstLoop = true;
+  AliParticleContainer * partCont = 0;
+  TIter nextPartCont(&fParticleCollArray);
+  while ((partCont = static_cast<AliParticleContainer*>(nextPartCont()))) {
+    if (firstLoop != true) {
+      particleContainerNames += "_";
+    }
+    else {
+      firstLoop = false;
+    }
+    particleContainerNames += partCont->GetArrayName();
+  }
+
+  fEmcalTracks = new TClonesArray("AliEmcalParticle");
+  fEmcalTracks->SetName(Form("EmcalTracks_%s", particleContainerNames.c_str()));
+
+  if (fCreateHisto){
+    fCellTrackMatchdEtadPhi = new TH2F("hCellTrackMatchdEtadPhi","hCellTrackMatchdEtadPhi",100,-0.04,0.04,100,-0.04,0.04);
+    fOutput->Add(fCellTrackMatchdEtadPhi);
+    fCellNTrackMatch = new TH2F("hCellNTrackMatch","hCellNTrackMatch",3,-0.5,2.5,500,0,20);
+    fOutput->Add(fCellNTrackMatch);
+    fCellTrackMatchEbefore = new TH1F("hCellTrackMatchEbefore","hCellTrackMatchEbefore",500,0,20);
+    fOutput->Add(fCellTrackMatchEbefore);
+    fCellTrackMatchEafter = new TH1F("hCellTrackMatchEafter","hCellTrackMatchEafter",500,0,20);
+    fOutput->Add(fCellTrackMatchEafter);
+  }
+
+
+}
+
+/**
+ * Called before the first event to initialize the correction.
+ */
+void AliEmcalCorrectionCellTrackMatcher::ExecOnce()
+{
+  fParticleContainerIndexMap.CopyMappingFrom(AliParticleContainer::GetEmcalContainerIndexMap(), fParticleCollArray);
+}
+
+/**
+ * Called for each event to process the event data.
+ */
+Bool_t AliEmcalCorrectionCellTrackMatcher::Run()
+{
+  AliEmcalCorrectionComponent::Run();
+
+  fEmcalTracks->Delete();
+  AliVTrack* track = 0;
+
+  AliParticleContainer * partCont = 0;
+  TIter nextPartCont(&fParticleCollArray);
+  while ((partCont = static_cast<AliParticleContainer*>(nextPartCont()))) {
+    auto partItCont = partCont->accepted_momentum();
+    for (AliParticleIterableMomentumContainer::iterator partIterator = partItCont.begin(); partIterator != partItCont.end(); ++partIterator) {
+      track = static_cast<AliVTrack *>(partIterator->second);
+      // Propagate the track
+      AliEMCALRecoUtils::ExtrapolateTrackToEMCalSurface(track, 440, 0.1396, 20, 0.35, kFALSE, kTRUE, kFALSE);
```

Comment:
You should check the cluster-track matcher for the cases where we actually have to propagate the tracks. For AODs, this is stored for Run2, which can save a lot of computation time. We recently introduced some improvements there. (This may have actually been in the RecoUtils, so maybe this is already accounted for. Take this as a suggestion to crosscheck this if you haven't already)